### PR TITLE
Upgrade libraries to spack-stack/1.5.0 to match the UFS versions

### PR DESCRIPTION
**Pull request by DavidHuber-NOAA**
## DESCRIPTION OF CHANGES: 
This upgrades the libraries to those currently used by the UFS as managed by spack-stack version 1.5.0 on Orion and Hera.  In particular, this upgrades netCDF-c to 4.9.2, netCDF-Fortran to 4.6.0, HDF5 to 1.14.0, ip to 4.3.0, w3emc to 2.10.0, and ESMF to 8.4.2.  The CI library versions were also updated to match these versions.

Lastly, during testing, it was determined that two tests needed additional resources.  chgres_cube with sigio used close to 60GB on Orion (increased memory request from 50GB to 75GB) and took just less than 15 minutes (increased time request to 20 minutes from 15) and cpld_gridgen at 025 took a little longer than 15 minutes on Orion (increased time request from 15 minutes to 20 minutes).

## TESTS CONDUCTED: 
If there are changes to the build or source code, the tests below must be conducted. Contact a repository manager if you need assistance.

- [x] Compile branch on all Tier 1 machines using Intel (Orion, Jet, Hera and WCOSS2).
- [x] Compile branch on Hera using GNU.
- [x] Compile branch in 'Debug' mode on WCOSS2.
- [x] Run unit tests locally on any Tier 1 machine.
- [x] Run relevant consistency tests locally on all Tier 1 machine.

Describe any additional tests performed.

## DEPENDENCIES:
None

## DOCUMENTATION:
N/A

## ISSUE: 
#859

### CODE REVIEW

**Comment by DavidHuber-NOAA** [comment]
I do not understand this CI error from the GCC ctests.  The same tests [passed successfully on commit](https://github.com/DavidHuber-NOAA/UFS_UTILS/actions/runs/6711537004/job/18239136043), but are failing only in this PR and it looks like it is running the same workflows.  Is there a difference between the tests run in the PR and those run on commit?

**Comment by GeorgeGayno-NOAA** [comment]
> I do not understand this CI error from the GCC ctests. The same tests [passed successfully on commit](https://github.com/DavidHuber-NOAA/UFS_UTILS/actions/runs/6711537004/job/18239136043), but are failing only in this PR and it looks like it is running the same workflows. Is there a difference between the tests run in the PR and those run on commit?

I see a lot of 'illegal' instructions. I can't explain it, but I have seen that error before. Rerunning the workflow usually fixes it.

**Comment by DavidHuber-NOAA** [comment]
@GeorgeGayno-NOAA Thanks!  I will likely push an update for Jet today, so I will let it run then.

**Comment by DavidHuber-NOAA** [comment]
Ran regression tests on all tier-1 platforms -- all passed with the only needed changes on Orion (time limits, memory) and Jet (different QOS).  The GCC CI ctests continue to fail, but I think that is because the environment needs to be built from scratch again, but the setup job does not clean the environment.  @GeorgeGayno-NOAA Any suggestions on how to clean the environment?

**Comment by GeorgeGayno-NOAA** [comment]
> Ran regression tests on all tier-1 platforms -- all passed with the only needed changes on Orion (time limits, memory) and Jet (different QOS). The GCC CI ctests continue to fail, but I think that is because the environment needs to be built from scratch again, but the setup job does not clean the environment. @GeorgeGayno-NOAA Any suggestions on how to clean the environment?

Rerunning that particular job usually fixes the problem.

@aerorahul - Any idea why the GCC test fails occasionally? Example:
https://github.com/ufs-community/UFS_UTILS/actions/runs/6722919365/job/18271879484?pr=866

**Comment by GeorgeGayno-NOAA** [comment]
@DavidHuber-NOAA The head of 'develop' no longer compiles on Orion. I am guessing I need to merge your PR next. When were the libraries used by UFS_UTILS removed?

**Comment by aerorahul** [comment]
> > Ran regression tests on all tier-1 platforms -- all passed with the only needed changes on Orion (time limits, memory) and Jet (different QOS). The GCC CI ctests continue to fail, but I think that is because the environment needs to be built from scratch again, but the setup job does not clean the environment. @GeorgeGayno-NOAA Any suggestions on how to clean the environment?
> 
> Rerunning that particular job usually fixes the problem.
> 
> @aerorahul - Any idea why the GCC test fails occasionally? Example: https://github.com/ufs-community/UFS_UTILS/actions/runs/6722919365/job/18271879484?pr=866

Not really.
It is possible the runner is glitchy.  
I am beginning to dislike the github runners for executing our codes.  
Using containers and running our codes in them on these runners is a safer bet IMO.

**Comment by DavidHuber-NOAA** [comment]
@GeorgeGayno-NOAA IIRC, EPIC does not actually own /work/noaa/epic-ps and the libraries that were installed there by EPIC were done so by accident.  I'm not sure when they were deleted exactly, but the 'correct' hpc-stack installation is here: `/work/noaa/epic/role-epic/contrib/orion/hpc-stack/intel-2022.1.2/modulefiles/stack`.

**Comment by DavidHuber-NOAA** [comment]
That said, I do not see netcdf/4.9.2 under that path, so it likely won't work for UFS_utils.

**Comment by GeorgeGayno-NOAA** [comment]
> That said, I do not see netcdf/4.9.2 under that path, so it likely won't work for UFS_utils.

Ok. Is your branch ready to merge?

**Comment by DavidHuber-NOAA** [comment]
@GeorgeGayno-NOAA Yes, I just verified there are no new commits in develop that need to be merged/tested.  It is ready to merge.

**Comment by GeorgeGayno-NOAA** [comment]
> @GeorgeGayno-NOAA Yes, I just verified there are no new commits in develop that need to be merged/tested. It is ready to merge.

Ok. To be safe I recommend you merge yesterday's updates from 'develop' to your branch.

**Comment by DavidHuber-NOAA** [comment]
@GeorgeGayno-NOAA Thanks for the heads up, I missed that.  I will rerun RTs on Jet today just to verify everything is still OK.

**Comment by DavidHuber-NOAA** [comment]
@GeorgeGayno-NOAA All regression tests passed on Jet after merging in develop.  I believe this PR is now ready to merge.

**Comment by GeorgeGayno-NOAA** [comment]
> @GeorgeGayno-NOAA All regression tests passed on Jet after merging in develop. I believe this PR is now ready to merge.

Several regression tests failed on Orion. I suspect the failures do not indicate a problem with your updates. I will confirm tomorrow. @DeniseWorthen - can you check the failed cpld_gridgen tests. My branch is here: `/work/noaa/stmp/ggayno/huber/UFS_UTILS/reg_tests/cpld_gridgen`

**Comment by DeniseWorthen** [comment]
@GeorgeGayno-NOAA Are other regression tests on orion passing, but the cpld_gridgen are not? Because I see no reason why these tests in particular would fail comparison. I appears that anything using ESMF weight generation is B4B different. I know when we updated the UWM to SS1.5.0 we had baseline changes. Do the cpld_gridgen baselines pass on hera but not on orion?

**Comment by DeniseWorthen** [comment]
EDIT: Disregard the below comment about location, I was looking at the hera location. 

Regarding non-B4B in UWM w/ the move to SS, I think those changes were all put down to the change in the MAPL library. I'm also remembering that there were issues w/ ESMF meshes (which this code uses) when they first began testing SS for UWM. Are we sure there are no build options required in UFS-UTILS that we're missing?

Another question---UWM uses this location as module file on orion:
```
prepend_path("MODULEPATH", "/work/noaa/epic/role-epic/spack-stack/orion/spack-stack-1.5.0/envs/unified-env/install/modulefiles/Core")
```

This doesn't seem to be the same location as in this PR.

**Comment by DavidHuber-NOAA** [comment]
@GeorgeGayno-NOAA @DeniseWorthen The regression tests failed for me on Orion yesterday as well -- I should have spoken to that.  Since the develop branch fails to build on Orion due to missing libraries (i.e. the recently deleted tree that was in `/work/noaa/epic-ps`).  However, the rt.sh script still attempts to launch the tests on Orion, but they obviously fail due to a lack of a comparison run.  This is why I opted to run the tests on Jet yesterday.

**Comment by GeorgeGayno-NOAA** [comment]
> @GeorgeGayno-NOAA Are other regression tests on orion passing, but the cpld_gridgen are not? Because I see no reason why these tests in particular would fail comparison. I appears that anything using ESMF weight generation is B4B different. I know when we updated the UWM to SS1.5.0 we had baseline changes. Do the cpld_gridgen baselines pass on hera but not on orion?

Other regression tests are failing on Orion, not just cpld_gridgen. A quick look at one of the chgres tests showed very small (ie. floating point) differences with the baseline. So, I don't think there is any real problem with these updates. But we should check.

**Comment by DeniseWorthen** [comment]
I also saw roundoff level differences in my spot check. I'm curious...the weights files that we just moved into the fix directory are now slightly different. Isn't the idea that they are fix files in conflict with needing new ones potentially each time SS updates?

**Comment by GeorgeGayno-NOAA** [comment]
> I also saw roundoff level differences in my spot check. I'm curious...the weights files that we just moved into the fix directory are now slightly different. Isn't the idea that they are fix files in conflict with needing new ones potentially each time SS updates?

Agree with your statement. Things change so rapidly at EMC now, that the concept of a 'fix' directory can have little meaning. 

If the differences are roundoff level, then I would not bother updating the 'fix' directory.

**Comment by DavidHuber-NOAA** [comment]
I reran regression tests on Hera after realizing that I was running develop only.  All regression tests failed, though consistency differences all appear to be at the roundoff level.  The regression tests were performed in `/scratch1/NCEPDEV/global/David.Huber/para/stmp/ufs-utils/UFS_UTILS`.

Comparing results against @GeorgeGayno-NOAA's tests revealed that only one test differs significantly between Orion and Hera: `grid_gen/c96.uniform`.  Consistency differences on Hera are no larger than `10^-7`.  On Orion, the grid_gen `c96.uniform` test is returning very large differences for the generated fix files.  All other `grid_gen` tests seem to be reasonable for both platforms.

**Comment by GeorgeGayno-NOAA** [comment]
> I reran regression tests on Hera after realizing that I was running develop only. All regression tests failed, though consistency differences all appear to be at the roundoff level. The regression tests were performed in `/scratch1/NCEPDEV/global/David.Huber/para/stmp/ufs-utils/UFS_UTILS`.
> 
> Comparing results against @GeorgeGayno-NOAA's tests revealed that only one test differs significantly between Orion and Hera: `grid_gen/c96.uniform`. Consistency differences on Hera are no larger than `10^-7`. On Orion, the grid_gen `c96.uniform` test is returning very large differences for the generated fix files. All other `grid_gen` tests seem to be reasonable for both platforms.

The differences on Orion can be explained. The previous merge changed results for the c96.uniform test. However, before I could update the baseline on Orion, the libraries changed so that 'develop' would not compile.

**Comment by DavidHuber-NOAA** [comment]
@GeorgeGayno-NOAA OK, that's good to know.  Would you like me to rerun regression tests on Jet as well?

**Comment by GeorgeGayno-NOAA** [comment]
> @GeorgeGayno-NOAA OK, that's good to know. Would you like me to rerun regression tests on Jet as well?

I have already run them on Jet: /lfs4/HFIP/emcda/George.Gayno/stmp/huber/UFS_UTILS. I have not checked results yet.

**Comment by GeorgeGayno-NOAA** [comment]
> > @GeorgeGayno-NOAA OK, that's good to know. Would you like me to rerun regression tests on Jet as well?
> 
> I have already run them on Jet: /lfs4/HFIP/emcda/George.Gayno/stmp/huber/UFS_UTILS. I have not checked results yet.

Fewer tests failed on Jet - just the snow2mdl tests and the cpld_gridgen tests. According to the log file, a few cpld_gridgen files differed from the baseline. However, when I manually checked them using `nccmp -dmfqS` there were no differences. I checked the snow2mdl file differences using Grads and differences were minor.

I don't see any problems on Jet.

File: modulefiles/build.jet.intel.lua [hunk 1]
```diff
@@ -1,64 +1,64 @@
...
+png_ver=os.getenv("png_ver") or "1.6.37"
+load(pathJoin("libpng", png_ver))
 
-g2_ver=os.getenv("g2_ver") or "3.4.5"
-load(pathJoin("g2", g2_ver))
+netcdf_c_ver=os.getenv("netcdf_c_ver") or "4.9.2"
+load(pathJoin("netcdf-c", netcdf_c_ver))
+
+netcdf_fortran_ver=os.getenv("netcdf_fortran_ver") or "4.6.0"
+load(pathJoin("netcdf-fortran", netcdf_fortran_ver))
```
Comment (GeorgeGayno-NOAA) [inline modulefiles/build.jet.intel.lua:54]: 'prod_util' is used by the 'gdas_init' utility. We need to add this back to this build script or add its module load to the 'gdas_init' driver script for jet.

**Comment by DavidHuber-NOAA** [comment]
@GeorgeGayno-NOAA Is there anything else that needs to happen for this PR?

File: modulefiles/build.jet.intel.lua (hunk 1 above)
Comment (GeorgeGayno-NOAA) [reply modulefiles/build.jet.intel.lua:54]: @DavidHuber-NOAA - one final thing before we can merge.

**Review by GeorgeGayno-NOAA — CHANGES_REQUESTED** [review]

File: modulefiles/build.jet.intel.lua (hunk 1 above)
Comment (DavidHuber-NOAA) [reply modulefiles/build.jet.intel.lua:54]: Thanks @GeorgeGayno-NOAA for the catch!  I ended up putting it in the `gdas_init` driver to match how Hera does it.  I tested this on Jet and successfully created ICs for 2023110106 in `/lfs4/HFIP/hfv3gfs/David.Huber/output`.

**Review by GeorgeGayno-NOAA — APPROVED** [review]

**Pull request merged**